### ISSUE TITLE
Enhancement: Build System, Cross Compiling, Android, Termux, Clang, CMake

### ISSUE THREAD

**Issue by Leif-W** [lcn2/calc#59]
**Is your feature request related to a problem? Please describe.**
Background, context.  I'm using Termux on an Android device to make Android apps in Android itself, for various reasons, including an exercise in learning, a "fun constraint", morbid curiousity, poverty, disability/accessibility, spiteful resentment of Google's HEAVY, PC-centric oppressive bias, closed-source, rottenly documented tools, as well as disdain for the ecosystem of closed-source, paid SaaS, like ReactNative/expo, and others.  Thus far, I've been able to build and install a basic "Hello World" app, with additional "Hello World" accessed via JNI to C.  Further, I cross compiled this trivial C library as a .so for both aarch64 and armv7a, assembled in a single .apk file (in directories lib/arm64-v8a and lib/armeabi-v7a, respectively), and tested on both hardware devices (aarch64 Android 12, armv7l ['el' not 'one'] Android 5).

This brings me to my next phase.  How do I cross compile libcalc as a .so on aarch64 for armv7a?  That's a lot to unpack.  Yes, the Termux project has compiled and packaged calc (thus libcalc) for Android, which is how I learned about calc.  I could just copy the .so files (I'm aware of the possible libcalc dependency on other .so's, not ignoring, just deferring concern).  However, it besides "cheating", it doesn't help if I'm trying to compile for armv7a.  Nor does it help if I am on a Linux or Windows PC (say AMD or Intel x64 or server CPU), either physically or remotely, and wish to cross compile libcalc.so for aarch64 and armv7a.

In Termux, when I compile calc from unmodified source using the minimum commands (just `make` or `make -f Makefile`), I see `gcc` is chosen, and there are lots of errors, and something ends in failure, yet there are some compiled binary executables and libs.

First problem, Google officially dropped gcc on Android years ago due primarily to GPL licensing and also cited project management issues as a (possibly bogus) excuse.  While there is an unofficial, quasi-maintained patch to gcc within Termux, itself an unofficial, sporadically maintained set of patches, that gcc can be a source of untold errors.  It seems to work fine, on simple programs, but who knows?  Clang is far more supported on Android in Termux.  Using `clang --target arm64-linux-android24` (redundant when I'm on the device) and `clang --target armv7a-linux-androideabi24` (with -fPIC -shared) seems to get the .so format that makes the Java/Android subsystem happy when calling System.loadLibrary().

I'm having difficulty knowing what command line options and/or variables to set, and/or if/what parts of the Makefile need to be modified to properly detect the Termux build environment, and how to build ONLY the libraries.

uname -s is Linux
uname -p is unknown
uname -m is aarch64
uname -o is Android (but this is stored in a variable named MINGW, ugh! hould be OS or OSNAME or something, an os-neutral variable name)

**Describe the solution you'd like**

I think I already have the ability to build the library, but ut seems to ONLY allow building shared library .so not only the .a libraries?  I don't think I need .a only right now, but it seemed odd to lack that granularity of control.

I'm not sure if the code checks can be run to test libraries only, or if that requires the full command line utility to be built in order to run the tests.  Separate lib-only tests would be nice, to isolate command line bugs from library bugs.  Maybe it doesn't yield enough benefit for the effort?  Clarity on this point?

I'd like to see, at a bare minimum, Android/Termux properly identified, clang chosen.

I'd really like to be able to specify cross-compilation, not assume that the current build-CPU arch is the only arch in the universe.

I'd really, really like to be able to specify multiple targets for cross compilation, run a single build command, and collect the binaries.  I can manage via scripting (shell or Python) as an interim workaround (loop for each target, clean/clobber, build, copy, next).

It'd be nice to have a CMake build system, to manage these sorts of things.  I'll let Santa, the Easter Bunny, and the Tooth Fairy know.  I'm sure they'll get right on that.  :o)  But seriously, has this been considered, discussed?  Is there a conversation somewhere that I may have missed?  If so, direct me there.  I'm curious about feasability, specific concerns, particularly regarding `complex dependency issues between `Makefile, Makefile.ship and custom/Makefile` which are currently blocking parallel builds, possibly blocking cross compilation and multi-target builds.  Does CMake offer easier access to these features?  As well as the boilerplate utility and OS testing, which may make a large portion of existing Makefile redundant?

Finally, I'd eventually like to have my Android calculator app build on Linux proper, as well as Windows under WSLv2, MinGW/MSYS (implies Cygwin, sort of), and maybe one day if possible Visual Studio Build Tools (command line C preprocessor, C compiler, library linker), though I'm not optimistic about ARM cross compilation with freely available tools from MS (they tend to charge money for advanced features and then rabidly vendor-lock in a limited subset of environments), but who knows, I may be ignorant, hence the exploration of possibility.

**Describe alternatives you've considered**
`make calc-dynamic-only LCC=clang CFLAGS="--target=arm64-linux-android24" BLD_TYPE=calc-dynamic-only`

```
forming have_limits.h
have_limits.h formed
forming have_ban_pragma.h
have_ban_pragma.h formed
forming charbit.h
charbit.h formed
forming have_unistd.h
have_unistd.h formed
forming have_stdlib.h
have_stdlib.h formed
forming longbits.h
longbits.h formed
forming align32.h
align32.h formed
forming have_string.h
have_string.h formed
forming args.h
args.h formed
forming calcerr.h
calcerr.h formed
forming conf.h
conf.h formed
forming endian_calc.h
endian_calc.h formed
forming have_fpos.h
have_fpos.h formed
forming have_posscl.h
have_posscl.h formed
forming have_fpos_pos.h
have_fpos_pos.h formed
forming have_offscl.h
have_offscl.h formed
forming have_newstr.h
have_newstr.h formed
forming have_memmv.h
have_memmv.h formed
forming fposval.h                                                           fposval.h formed
forming have_const.h                                                        have_const.h formed                                                         forming have_getpgid.h
have_getpgid.h formed
forming have_getprid.h
have_getprid.h formed
forming have_getsid.h
have_getsid.h formed
forming have_gettime.h
have_gettime.h formed
forming have_rusage.h
have_rusage.h formed
forming have_strdup.h
have_strdup.h formed
forming have_strlcat.h
have_strlcat.h formed
forming have_strlcpy.h
have_strlcpy.h formed
forming have_times.h
have_times.h formed
forming have_uid_t.h
have_uid_t.h formed
forming have_unused.h
have_unused.h formed
forming have_urandom.h
have_urandom.h formed
forming have_ustat.h
have_ustat.h formed
forming terminal.h
terminal.h formed
forming have_environ.h
have_environ.h formed
forming have_arc4random.h
have_arc4random.h formed
forming calcerr.c
calcerr.c formed
rm -f custom/.all
cd custom; make -f Makefile ALLOW_CUSTOM="-DCUSTOM" AR="ar" ARCH_CFLAGS="" AWK="awk" BINDIR="/usr/bin" BLD_TYPE="calc-dynamic-only" CALC_INCDIR="/usr/include/calc" CALC_SHAREDIR="/usr/share/calc" CAT="cat" CC=" clang " CCBAN="-UUNBAN" CCERR="" CCMISC="" CCOPT="-O3 -g3" CCWARN="-Wall -Wextra -pedantic" CC_SHARE="-fPIC" CFLAGS="--target=arm64-linux-android24 -I.." CHMOD="chmod" CMP="cmp" CO="co" COMMON_ADD="" COMMON_CFLAGS="-DCALC_SRC -DCUSTOM -Wall -Wextra -pedantic     -I.." COMMON_LDFLAGS=" " CP="cp" CUSTOMCALDIR="/usr/share/calc/custom" CUSTOMHELPDIR="/usr/share/calc/custhelp" CUSTOMINCDIR="/usr/include/calc/custom" DEBUG="-O3 -g3" DEFAULT_LIB_INSTALL_PATH="/data/data/com.termux/files/home/dev/calc-2.14.1.0:/lib:/usr/lib:/usr/lib:/usr/local/lib" DIFF="diff" E="2>/dev/null" FMT="fmt" GREP="egrep" H="@" HELPDIR="/usr/share/calc/help" ICFLAGS="-DCALC_SRC -DCUSTOM -Wall -Wextra -pedantic     -UUNBAN -fPIC -I.." ILDFLAGS=" " INCDIR="/usr/include" LANG="C" LCC="clang" LDCONFIG="ldconfig" LDFLAGS="  " LD_SHARE=""-Wl,-rpath,/data/data/com.termux/files/home/dev/calc-2.14.1.0:/lib:/usr/lib:/usr/lib:/usr/local/lib" "-Wl,-rpath-link,/data/data/com.termux/files/home/dev/calc-2.14.1.0:/lib:/usr/lib:/usr/lib:/usr/local/lib"" LIBCUSTCALC_SHLIB="-shared "-Wl,-soname,libcustcalc.so.2.14.1.0"" LIBDIR="/usr/lib" LN="ln" LS="ls" MAKE="make" MAKEDEPEND="makedepend" MAKE_FILE=Makefile MKDIR="mkdir" MV="mv" MINGW="Android" PREFIX="/usr/local" PURIFY="" Q="@" RANLIB="ranlib" RM="rm" RMDIR="rmdir" S=">/dev/null 2>&1" SCRIPTDIR="/usr/bin/cscript" SED="sed" SHELL="/bin/bash" SORT="sort" T="" TAIL="tail" TOUCH="touch" TRUE="true" V="@:" VERSION="2.14.1.0" WNO_IMPLICT="-Wno-implicit" WNO_ERROR_LONG_LONG="-Wno-error=long-long" WNO_LONG_LONG="-Wno-long-long" target="Linux" all
make[1]: Entering directory '/data/data/com.termux/files/home/dev/calc-2.14.1.0/custom'
clang  --target=arm64-linux-android24 -I..   -c -o custtbl.o custtbl.c
In file included from custtbl.c:30:
In file included from ../value.h:42:                                        /data/data/com.termux/files/usr/include/calc/decl.h:29:10: fatal error: 'calc/have_const.h' file not found
#include "calc/have_const.h"
         ^~~~~~~~~~~~~~~~~~~
1 error generated.
make[1]: *** [<builtin>: custtbl.o] Error 1
make[1]: Leaving directory '/data/data/com.termux/files/home/dev/calc-2.14.1.0/custom'
make: *** [Makefile:4373: custom/.all] Error 2
```

**Additional context**
I used `bc` for nearly 3 decades.  Then I found that it couldn't do decimal exponentiation.  Then I found GMP but it has no CLI.  Then I found calc via Termux package system, and it has a CLI, a library, and handles decimal exponentiation.  Used this for a couple years.  Then I had the (dumb, crazy, fun) idea to make a calculator app, after my old propietary app started spying.  Existing Android build processes depend way too much on bloated, buggy, exponentially complex, proprietary nonsense.  So I decided to build on Android in Termux using only bash, vim, openjdk's javac, jar, and keytool, android.jar, and whatever other build tools are necessary, like aapt2, d8 (DEX v8), 7z (or Info-Zip's zip would work), zipalign, apksigner, later, clang and llvm-strip, eventually git to publish on GitHub.  I looked again at GMP and it was highly recommended by a friend, but the exponentiation functions only handle integers, meaning a fair bit of cooking the decomal into reduced rational form and doing many steps, common tasks which belong in a library as utility wrappers, which GMP is uninterested in.  Also, I find that some of GMPs other decimal functions provide little or no better accuracy than IEEE 754 with its rounding errors.  Hence the need to also use the MPFR library for all floating points.  Ok.  But the current release is 4.x.  There's a Java binding library for MPFR but only v3.x.  So I'd have to do the Android to JNI to C wrappers myself, straightforward, hopefully.  MPFR lists `aarch64-unknown-linux-gnu (gcc 8.5.0 and 12.2.0, clang 15.0.4 on Termux/Android, tcc mob branch);` as the very first supported platform, with both gcc and clang testing.  It's nice to be placed as a first class platform, rather than an annoyance to be dismissed.  My phone and old desktop PC are nearly equal in power.  If my truck breaks down, I can use my phone to get help.  I can't make a phone call by telepathic link to my desktop and then VOIP from my desktop.  Can only afford one upgrade.  Needed to delay desktop/laptop upgrade.  Also, disability randomly makes sitting at desk impossible, sometimes for weeks at a time.  It's easier to hold a 4oz phone in my hand while reclining, than an 80lb computer (including all peripherals).  Sure, it'd be nice to afford a big enough room to live in and afford a large reclining chair with computer, monitor, and keyboard/mouse mounts, but see constraint above regarding computer upgrade.  Even a laptop would honestly be very difficult to use.  A tablet might be better, but can't call for help if I fall, get stuck, or car breaks down, or illness temporarily impairs ability to get to landline phone.  Hence why I ended up on Android as my "primary" platform.  All of that's no reason to stop learning and trying to build things.  :o)

**Comment by lcn2** [comment]
Interesting issue: although going into a new build system might take a while for someone to get around to trying and testing, not to mention all of the existing code base that uses the current `make(1)` that could be negatively impacted.

I wonder if the better thing to do is to focus on what seems to be a bug:

```
cd custom; make -f Makefile ALLOW_CUSTOM="-DCUSTOM" AR="ar" ARCH_CFLAGS="" AWK="awk" BINDIR="/usr/bin" BLD_TYPE="calc-dynamic-only" CALC_INCDIR="/usr/include/calc" CALC_SHAREDIR="/usr/share/calc" CAT="cat" CC=" clang " CCBAN="-UUNBAN" CCERR="" CCMISC="" CCOPT="-O3 -g3" CCWARN="-Wall -Wextra -pedantic" CC_SHARE="-fPIC" CFLAGS="--target=arm64-linux-android24 -I.." CHMOD="chmod" CMP="cmp" CO="co" COMMON_ADD="" COMMON_CFLAGS="-DCALC_SRC -DCUSTOM -Wall -Wextra -pedantic     -I.." COMMON_LDFLAGS=" " CP="cp" CUSTOMCALDIR="/usr/share/calc/custom" CUSTOMHELPDIR="/usr/share/calc/custhelp" CUSTOMINCDIR="/usr/include/calc/custom" DEBUG="-O3 -g3" DEFAULT_LIB_INSTALL_PATH="/data/data/com.termux/files/home/dev/calc-2.14.1.0:/lib:/usr/lib:/usr/lib:/usr/local/lib" DIFF="diff" E="2>/dev/null" FMT="fmt" GREP="egrep" H="@" HELPDIR="/usr/share/calc/help" ICFLAGS="-DCALC_SRC -DCUSTOM -Wall -Wextra -pedantic     -UUNBAN -fPIC -I.." ILDFLAGS=" " INCDIR="/usr/include" LANG="C" LCC="clang" LDCONFIG="ldconfig" LDFLAGS="  " LD_SHARE=""-Wl,-rpath,/data/data/com.termux/files/home/dev/calc-2.14.1.0:/lib:/usr/lib:/usr/lib:/usr/local/lib" "-Wl,-rpath-link,/data/data/com.termux/files/home/dev/calc-2.14.1.0:/lib:/usr/lib:/usr/lib:/usr/local/lib"" LIBCUSTCALC_SHLIB="-shared "-Wl,-soname,libcustcalc.so.2.14.1.0"" LIBDIR="/usr/lib" LN="ln" LS="ls" MAKE="make" MAKEDEPEND="makedepend" MAKE_FILE=Makefile MKDIR="mkdir" MV="mv" MINGW="Android" PREFIX="/usr/local" PURIFY="" Q="@" RANLIB="ranlib" RM="rm" RMDIR="rmdir" S=">/dev/null 2>&1" SCRIPTDIR="/usr/bin/cscript" SED="sed" SHELL="/bin/bash" SORT="sort" T="" TAIL="tail" TOUCH="touch" TRUE="true" V="@:" VERSION="2.14.1.0" WNO_IMPLICT="-Wno-implicit" WNO_ERROR_LONG_LONG="-Wno-error=long-long" WNO_LONG_LONG="-Wno-long-long" target="Linux" all
make[1]: Entering directory '/data/data/com.termux/files/home/dev/calc-2.14.1.0/custom'
clang  --target=arm64-linux-android24 -I..   -c -o custtbl.o custtbl.c
In file included from custtbl.c:30:
In file included from ../value.h:42:                                        /data/data/com.termux/files/usr/include/calc/decl.h:29:10: fatal error: 'calc/have_const.h' file not found
#include "calc/have_const.h"
         ^~~~~~~~~~~~~~~~~~~
1 error generated.
make[1]: *** [<builtin>: custtbl.o] Error 1
make[1]: Leaving directory '/data/data/com.termux/files/home/dev/calc-2.14.1.0/custom'
make: *** [Makefile:4373: custom/.all] Error 2
```

Perhaps?

**Comment by Leif-W** [comment]
Thanks for the reply.  Whatever the path of least resistence to get to my ultimate goal is all that matters.

First, before declaring a bug, I want to be sure I am correctly specifying my needs via the existing means offered by the current build system.  To the best of my ability, I have determined that my Android/Termux can generate binary shared object library files in a format that makes the Android app, and underlying Java and JNI systems "happy".

I need to use `clang -target arm64-linux-android24 -shared -fPIC` and `clang -target armv7a-linux-androideabi24 -shared -fPIC`, respectively, for 32 and 64-bit ARM CPUs.

First, how do I correctly specify these?  clang compiler, specific target for cross compilation, and -shared and -fPIC (if necessary, as they seem to be specified by default for gcc as well).

Second, how to patch the make file to correctly identify Linux/aarch64/Android/Termux?  This is independent of any Android/JNI stuff.  This effects merely building and using the CLI as I would on any other system.  The default target of aarch64/arm64/arm64-v8a need not be explicitly specified in this case.

Thirdly, if I am specifying custom build settings correctly, and if the make build system iddntifies my platform correctly and selects the correct compiler and options, and bugs like tests and finding headers and code warnings still occur, then we address those each in turn.

Fourthly, I forgot to mention a deferred concern, if the arm64 platform is verified supported, does it rely on hand coded assembly, are there compiler quirks that need addressing with flags?  So ideally, we get the code to compile without error, then run test suite to verify correct functionality.  If there are calculation, rounding, other runtime weirdness, we address that last?

As for setting up a test environment, if that would be helpful and if interested, it's relatively easy if you have a modern Android phone.  If not interested, skip paragraph.  But it's also possible to use an Android Studio or other emulator on a PC.  Install F-Droid (app repo), and [Termux](https://termux.dev/).  Then try to run the commands you need.  If missing, a script is automatically run to guess and suggest packages to install.  `pkg se <name>` will also search possible matching packages and let you know if it is installed.  `pkg` seems to be a wrapper for `apt`, and uses a Debian style `etc/apt/sources.list` config file.  Tools should be available with the default repo source, `deb https://packages.termux.dev/apt/termux-main stable main`.

Anyways, please advise on steps 1-4 above, if they are reasonable and correct, or suggest how best or preferred to proceed.  As for my end, I can fork and clone the calc repo with git running on the device itself.  I'm not as familiar with the git workflow and syntax to create a test branch and make a pull request, so there may be initial questions or mistakes.  Otherwise, I can also run `diff` with any options you desire, such as a unified diff, or with colors, discrete patches in a 7z or zip file, etc, as a fallback and last resort, if my Git/GitHub proficiency is problematic.

**Comment by lcn2** [comment]
While this may not solve your problem, please download and try [calc v2.14.1.1](https://github.com/lcn2/calc/releases/tag/v2.14.1.1), @Leif-W 

If you still run into problems, please attach, as requested in the bug report template:

```
IMPORTANT: Please attach the output from `make debug`.
```

Thanks.

**Comment by Leif-W** [comment]
Not sure if I missed the template instruction for make debug, sorry.

I did a git clone, which is only a few commits ahead of 2.14.1.1.  Ahh, I see the response was edited to refer to a template commit.

The make debug command chooses gcc which does not address my concern as the system does not detect Android/Termux platform which uses clang.  Nor does it address cross-compilation.

Github doesn't seem to provide any way to attach a 0.5 MB txt file with .out format  Changed to a more descriptive name and a .log extension, which worked ok.  Suggest that change to use .log or .txt filename extension in tge debug rule and template instruction.

[calc-make-debug.log](https://github.com/lcn2/calc/files/10114134/calc-make-debug.log)
.

It seems to build and pass tests using gcc, but throws a lot of warnings, probably due to warning levels, which I'm not sure whst can safely be ignored.  This works as I have gcc installed and gcc seems to be in a current working state, but there's no guarantee of this, as clang is the primary supported compiler.

**Comment by lcn2** [comment]
> Github doesn't seem to provide any way to attach a 0.5 MB txt file with .out format. Changed to a more descriptive name and a .log extension, which worked ok. Suggest that change to use .log or .txt filename extension in the debug rule and template instruction.
> 
> [calc-make-debug.log](https://github.com/lcn2/calc/files/10114134/calc-make-debug.log)

Good to note.  We all change the `make debug` to produce the file `debug.YYYYMMDD.HHMMSS.txt` in the next release.

**Comment by lcn2** [comment]
> It seems to build and pass tests using gcc, but throws a lot of warnings, probably due to warning levels, which I'm not sure what can safely be ignored. This works as I have gcc installed and gcc seems to be in a current working state, but there's no guarantee of this, as clang is the primary supported compiler.

Please give calc v2.14.1.2 a try, @Leif-W 

That version was released to address some issues related to musl libc and the gentoo Linux folks.  While it might not fix all of your issues, it should help with a few.

In the mean time we will look at the [calc-make-debug.log](https://github.com/lcn2/calc/files/10114134/calc-make-debug.log) you submitted.

A revised `debug.log` attachment for the current top of the master branch would be appreciated.

In the mean time, you could help us create better defaults for  Linux/aarch64/Android/Termux by giving us the output of these commands in that system:

```sh
$ uname -s

$ uname -p

$ uname -m

$ uname -o

$ uname -a
```

Thanks!

**Comment by Leif-W** [comment]
Might it be possible to include uname output in the make debug log?

`uname -s`
Linux

`uname -p`
unknown

`uname -m`
aarch64

`uname -o`
Android

`uname -a # I see it's already included in debug`
Linux localhost 4.19.113-24906469 #1 SMP PREEMPT Wed Oct 19 20:13:34 KST 2022 aarch64 Android

Other env variables which may be of interest:

```
BASH_VERSION=5.2.12(1)-release
TERMUX_VERSION=0.118.0 # existence implies Termux
PREFIX=/data/data/com.termux/files/usr # Termux should be here
```

People may also use any of several other shells, but Bash is default.

Fresh clone, `make debug`.

[debug.log](https://github.com/lcn2/calc/files/10122435/debug.log)

I also note, although make detects a gcc command, it''s really clang.  Although many command line options are identical, some may not be, and behavior may differ.

```
=-=-= LCC=gcc =-=-=                                                               
=-=-= output of ${LCC} -v follows =-=-=
clang version 15.0.5
Target: aarch64-unknown-linux-android24
```

Lokking at each warning:

- have_getprid - We don't have in Termux, I guess the test found that out, ignore?  We have a getgrid but probably unrelated.
- have_uid_t / geteuid - The function prototype is in unistd.h but maybe it's broken in Termux?  Ask them.  Termux is by default single user only, enforced by Android, so user and group are nearly meaningless, except for Android to enforce restrictions on our access, unless rooting device, which can vary from rarely trivial, to some non-trivial, to most devices being nearly impossible.  So it's not surprising if the user, group id stuff is crufty like this.
- func.c - 202:22 function declaration without a prototype is deprecated in all versions of C for both numfunc and valfunc, with subsequent warnings 'passing arguments to a function without a prototype is deprecated in all versions of C and is not supported in C2x'
- opcodes.c - similar warnings as func.c
- symbol.c - variable 'count' set but not used
- calc.c - variable 'unusedint' set but not used // self descriptive
- egrep: warning: egrep is obsolescent; using grep -E
- funclist.c - similar error to func.c numfunc/valfunc.  Also, argc and argv unused.
- Another egrep usage.
- <times.h> - error, not found, but we do have <sys/times.h>
- <ustat.h> - Missing header files, but we have the ustat man page only, which informs deprecation, and advises <statfs.h> which we have both headers for all supported Android archs and the man pages.
- 8703 - This test may have had errors?  I'm not sure what the output means, but it stood out as more verbose than others.  May be as intended.

That's about everything my cursory review of the debug log yielded, case-insensitive searching for "warn" and "error" keywords.

**Comment by lcn2** [comment]
Thanks for the very detailed report @Leif-W 

We will look over your information in detail.

**Comment by lcn2** [comment]
Just to let you know, @Leif-W we are experiencing Internet connectivity problems that may last through Tuesday, for some 24 hour value of Tuesday :), and thus it may be several days before we can address your recent information. 

Thank you for updating us, and we hope to, when Internet services restored, get back to your issue soon. Sorry (tm Canada 🇨🇦) for the inconvenience and thanks for your patience.

**Comment by Leif-W** [comment]
No problem @lcn2 , thanks for letting me know.

**Comment by lcn2** [comment]
Just an update, @Leif-W,

We will make this a priority .. although right now we have a regional internet service outage .. this reply is being send thru a satellite phone ($$$-ive). Nevertheless we wanted to let you know that we will address this issue, work offline on the matter, and when our server is restored (in a few days is the ETA) we will get back to you.

**Comment by Leif-W** [comment]
No rush from me, no stress to you, @lcn2 .  Life is life, things happen.  Understood.  Stay safe.

**Comment by lcn2** [comment]
Hello @Leif-W,

In regards to your [comment 1332064309](https://github.com/lcn2/calc/issues/59#issuecomment-1332064309) above, please check out the top of the master branch.

In particular, please try to compile again and let us know if the warnings you found have been addressed.

BTW:  These items are not fixed because they are not errors:

```
- <times.h> - error, not found, but we do have <sys/times.h>
```

This is not an error.  When forming have_times.h, calc looks for <times.h> and when it does not find it (the compile fails), the have_times.h is written to undefine HAVE_TIMES_H.

You will also notice that HAVE_SYS_TIMES_H is defined in have_times.h, inciting that you have <sys/times.h>.

```
- <ustat.h> - Missing header files, but we have the ustat man page only, which informs deprecation, and advises <statfs.h> which we have both headers for all supported Android archs and the man pages.
```

This is not an error.  When forming have_ustat.h, calc looks for <ustat.h> and when it does not find it (the compile fails), the have_ustat.h  written to undefine HAVE_USTAT.

There is new support for using states().  See the file have_statfs.h after calc is compiled.

**Comment by Leif-W** [comment]
I appreciate the time taken to consider this issue, comment, and modify code.

I fetched then pulled local (on phone, in Termux) git repo, status showed up to date, no locally modified files, good clean build.

Did a simple `make` and `make check` and `make chk`.  I didn't see any errors at all.  Wasn't sure what the output of `make check` meant, thus I did `make chk`, which simply said `chk OK`.  I did a `make clobber`.  (Is `make clean` or `make clobber` the preferred way to delete all build files and start a fresh build?)  Then I `make calc-dynamic-only BLD_TYPE=calc-dynamic-only`, and again, `make check` and `make chk`.  Again, no errors.

That's fine for a native armv8a / arm64 / aarch64 build.

But some things I noticed.

1. There's no Android and/or Termux detection (similar to Darwin), so the default `gcc` compiler is chosen, which does not exist by default on Termux, and is possibly broken or unsupported.  Android uses Clang.  Termux included a symbolic link from `gcc -> clang-15`.  Not sure if all of gcc's arguments are understood by clang, or if they function the same.  Other Android distros maybe don't have a symlink for gcc?  Not sure if this is intended as an exercise for the user, and if so, where to put the changes, although I guess `Makefile.local` is the correct place to override `Makefile` defaults?

2. Not sure how to achieve cross compilation, i.e. for an armv7a.  To generate an .so usable by an app via JNI, clang needs `--target arm64-linux-android24` for arm64, and for armv7a it needs `--target armv7a-linux-androideabi24`.  Is this something I must modify the Makefile.local to achieve, or is there a way to specify a cross compile on the command line by setting an environment variable for make to pick up, or what is the preferred method?  For example, maybe I am in Windows, and use MinGW to compile the mobile app?  Or, I use WSLv2 with Ubuntu or Debian to compile the mobile app?  In this case, I need to specify both targets, as both are non-native.  I need to test this further, later.  A bit tired now.

3. Just a side note.  I did an `llvm -s *.so.*` to strip all symbols rom the dynamic build, and `make chk` still came back with `chk OK`.  This greatly reduced the file size, from over 2 MB to under 1 MB, which is handy to keep a mobile app as small as possible.  Are there typically any unforeseen problems with this?

Generally, improved, no errors with native build.

**Comment by lcn2** [comment]
Thank you for the reply and confirmation that the current calc code is an improvement.

> 1. There's no Android and/or Termux detection (similar to Darwin), so the default `gcc` compiler is chosen, which does not exist by default on Termux, and is possibly broken or unsupported.  Android uses Clang.  Termux included a symbolic link from `gcc -> clang-15`.  Not sure if all of gcc's arguments are understood by clang, or if they function the same.  Other Android distros maybe don't have a symlink for gcc?  Not sure if this is intended as an exercise for the user, and if so, where to put the changes, although I guess `Makefile.local` is the correct place to override `Makefile` defaults?

We would be happy .. yea pleased to add a detection of "Android and/or Termux" as part of the native calc build.

Can you think of some way to detect your "Android and/or Termux" instance?

Early on in the Makefile we have:

```make
ifeq ($(target),)
target=$(shell uname -s 2>/dev/null)
endif
ifeq ($(arch),)
arch=$(shell uname -p 2>/dev/null)
endif
ifeq ($(hardware),)
hardware=$(shell uname -m 2>/dev/null)
endif
```

So perhaps some variables can be set so that later on, there is a "if Android and/or Termux" section that would in turn add to / replace / set variables suitable for the "Android and/or Termux" platform?

Or would you suggest that a simple private addition of a `Makefile.local` is good enough?

In the mean time we will declare, @Leif-W this issue to be closed.  Even so, if it would help you compile directly on such platforms by the some sort of unique Makefile test, we would be happy to consider doing so.

**Comment by Leif-W** [comment]
I wouldn't have closed it yet as there were several unanswered questions about cross-compilation, and some untested things.

As for detecting the Android operating system, as mentioned above, `uname -o` is `Android`, but that value is stored in a poorly named variable, `MINGW`, which is used for a single specific OS:

```
ifeq ($(MINGW),)
MINGW=$(shell uname -o 2>/dev/null)
ifeq ($(MINGW),Cygwin)
```

Rather, `uname -o` should more properly be called next to the other `uname` calls in the top level scope, and the value stored in a more generic variable name, such as `os` or better, `osname` or something unique, as `os` greps too many other things.

Rather, something like this?

```
ifeq ($(OSNAME),)
OSNAME=$(shell uname -o 2>/dev/null)
ifeq ($(OSNAME),Cygwin)
```

Then change it in all the other var dump strings in Makefile and similarly modify Makefile.simple.

I can go test this a bit later.

However, `uname -o` only tells us if we're in Android, but we may not necessarily be in Termux.  But if we're in Termux, we're always in Android.  Since Android can run on aarch64, armv7a, and x64 (maybe some old x86?), we can't use `uname -m` in this case.  Nor can `uname -p` help us, as it is `unknown`.  Nor does `uname -s` help, as `Linux` could mean anything, from a server, to a Raspberry Pi or something.

But if we're in Android, we can look at two environment variables for hints:

`$PREFIX` will always have `termux` in it, always (I think?) `/data/data/com.termux/files/usr`.

`$TERMUX_VERSION` will always exist and be set to some version number, irrelevent for us, so long as it's set to something.

Taking this together, we may do a section such as this:

```ifeq ($(OSNAME),Android)
ifeq ($(TERMUX),)
TERMUX=$(shell echo -n $TERMUX_VERSION)
# ... close ifeqs, set other things for Android, etc, then later ...
ifneq ($TERMUX),)
# Set Termux-specific things, run Termux-specific tests, etc?
```

I can make an attempt to patch this and submit a diff for review, discussion, and refinement.

I was able to set `CC=clang` in `Makefile.local` to use clang instead of gcc.  That would go in the Android section, possibly in the Termux section.  Not sure if CC is correct, or LCC is the more correct variable, to match the default Makefile.

The next problem, however, seems to be trying to build with a `Makefile.local` variable `CFLAGS+=-target armv7a-linux-androideabi24` and `LDFLAGS` set to append the same.  The FULL typedef is set to 64 bits, but the generated code is a 32 bit arch.  There seems to be no test to detect if a non-native target is used, to generate the bits testing code against the cross compiled target rather than the native cpu.  Or maybe I am doing this wrong?  I can try to manually set all the BITS variables to 32 a bit later, but I am curious about how to have this detection done automatically.  At the worst case scenario, I can manually specify all the bits variables.  But this is tedious.  I am trying to build shared library .so files for multiple targets.  I shouldn't have to edit any files.  I should ideally just have my build script run a command with a single different argument, the -target argument, for each desired target.  If I have to create multiple Makefile.targets, and have my script copy those into calc source tree to Makefile.local, and then run commands, so be it, but it feels like I'm doing something wrong by shuffling files around for every app, rather than handle the logic in one place, cross-compiling ability in calc.  I'll experiment further with this later this weekend.  If I get some ideas to work, I'll submit a diff/patch or pull request for possible comment or refinement.

**Comment by lcn2** [comment]
Hello @Leif-W,

This issue was reopened.

We also made a modification to the Makefile to rename MINGW to OSNAME.

We will await your changes for your Android / termux changes.

**Comment by Leif-W** [comment]
Thank you for the reopen.  I am still exploring this.

TL;DR:

- Added missing CFLAGS
- Disabled readline
- 3 warnings (1x fposval.h, 2x hist.c)
- binaries created
- make chk OK (maybe?)
- need to cleanup changes
- need to test further
- idea for enhancement: lib-only target

First issue was that when I specified the clang-specific `-target armv7a-linux-androideabi24` option in `CFLAGS` and `LDFLAGS`, as well as blindly setting all `*_BITS` values to 32 (not sure if there are any 16 or 64 bit in some cases?) a few things compiled as the 32-bit armv7a, yet others failed with an error about aarch64.  That told me several things were being compiled without CFLAGS or linked without LDFLAGS.  Looking closely, I saw it was mostly (entirely?) `${CC}` or `${LCC}` calls.  So I went through the top level `Makefile` and several sub-level `Makefile`s, searched for `CC}` [sic] (quick and dirty way to find ${LCC} or ${CC} without regexp), and added `${CFLAGS}`.  If there was already an `ICFLAGS` or `ILDFLAGS`, then I made no change.  Things mostly compiled then, but died because of some rule that modified some other files, `.bak` file errors, etc.  I just did a `make clobber`, or `rm Makefile.simple.bak`, or whatever, and ran the make again.  This time, `make` completed without failing.  There is a warning (error?) with `fposval`:

```
forming fposval.h                                                           WARNING!! ./fposval failed, using fposval.h.def
```

And also, two compiler warnings:

```
clang  -DCALC_SRC -DCUSTOM -Wall -Wextra -pedantic     -UUNBAN -fPIC -O3 -g3 -target armv7a-linux-androideabi24    \
    -c hist.c                                                               hist.c:101:24: warning: a function declaration without a prototype is deprecated in all versions of C [-Wstrict-prototypes]                             typedef void (*FUNCPTR)();
                       ^                                                                            void                                                hist.c:656:15: warning: passing arguments to a function without a prototype is deprecated in all versions of C and is not supported in C2x [-Wdeprecated-non-prototype]                                                                             (*ent->func)(key);                                                                      ^                                               2 warnings generated.
```

There was also an error about the `readline`, `ncurses`, and `history` libraries not being found.  I only want to bbuild the lib, not the command-line front-end, thus no need for these.  This may be a separate issue and enhancement, to have an option to build the lib only, and ignore these 3 dependencies.  Anyways, for testing, I just set `USE_READLINE`, `READLINE_LIB`, `READLINE_EXTRAS`, and `READLINE_INCLUDE` all to empty string in `Makefile` (should do in Makefile.local later, or in an `ifeq` block if building a lib-only target).  But the `hist.c` warnings/errors indicate that either I did not properly disable readline, or hist.c doesn't cleanly handle history being disabled, or something.

For now, I just ignore the 3 warnings, although the fposval.h warning concerns me moe than the hist.c ...

Also, somehow `make chk` returned `OK`.  That means the armv7a binaries ran on armv8a/aarch64, so the newer CPU is either backwards compatible, which I hadn't realized, or something else going on?  I've still got to learn how to write the headless C wrapper functions and the JNI wrappers and package in an .apk file to test on an actual armv7a device to verify a successful cross-compile.

**Comment by lcn2** [comment]
Hello @Leif-W ,

In the latest top of the master branch:

We fixed `hist.c` pointers to function warnings.

We fixed when USE_READLINE, READLINE_LIB, READLINE_EXTRAS, and READLINE_INCLUDE are set to empty (disable use of the GNU-readline facility).

We had been asked to address the case with the both GNU-readline and ncurses were disabled / not used.  The above does that AND it should help in your configuration where you are also not using GNU-readline.

**Comment by lcn2** [comment]
Hello @Leif-W ,

> For now, I just ignore the 3 warnings, although the fposval.h warning concerns me moe than the hist.c ...

The warnings should go away, but we agree with you about your fposval.h warning concern.

After you do a:

```sh
make clobber all chk
```

What does this command produce:

```sh
rm -f fposval.h; make fposval.h Q= S=
```

**Comment by lcn2** [comment]
> Also, somehow make chk returned OK. That means the armv7a binaries ran on armv8a/aarch64, so the newer CPU is either backwards compatible, which I hadn't realized, or something else going on? I've still got to learn how to write the headless C wrapper functions and the JNI wrappers and package in an .apk file to test on an actual armv7a device to verify a successful cross-compile.

When `make chk` returns OK, the the core of the calc language processing and computation passes all calc regression tests.  This is a good sign that calc is working well and that things like `regress.cal` are passing a rigorous set regression tests.

BTW: The more verbose check is:

```
make check
```

This does the same thing as `make chk` except it is more verbose.

**Comment by lcn2** [comment]
We are assuming that all is well @Leif-W and will now close this enhancement.
